libs/shared/providers/src/txButton/components/TxButton.tsx

```diff
@@ -0,0 +1,209 @@
+import { useEffect, useState } from 'react';
+
+import { capitalize } from '@mui/material';
+import { isUserRejected } from '@origin/shared/utils';
+import { usePrevious } from '@react-hookz/web';
+import { useIntl } from 'react-intl';
+import {
+  useAccount,
+  useSimulateContract,
+  useWaitForTransactionReceipt,
+  useWriteContract,
+} from 'wagmi';
+
+import { ConnectedButton } from '../../wagmi';
+
+import type { TransactionReceipt } from 'viem';
+
+import type {
+  ConnectedButtonProps,
+  WriteTransactionCallbacks,
+  WriteTransactionParameters,
+} from '../../wagmi';
+
+export type TxButtonProps = {
+  label?: string;
+  waitingSignatureLabel?: string;
+  waitingTxLabel?: string;
+  params: WriteTransactionParameters;
+  callbacks?: WriteTransactionCallbacks;
+} & Omit<
+  ConnectedButtonProps,
+  'onClick' | 'value' | 'children' | 'targetChainId'
+>;
+
+export const TxButton = ({
+  label,
+  waitingSignatureLabel,
+  waitingTxLabel,
+  params,
+  callbacks,
+  disabled,
+  ...rest
+}: TxButtonProps) => {
+  const intl = useIntl();
+  const { isConnected, chain } = useAccount();
+  const [txButtonState, setTxButtonState] = useState({
+    label: label ?? capitalize(params.functionName),
+    disabled,
+  });
+  const { data: simulateData, error: simulateError } = useSimulateContract({
+    address: params.contract.address,
+    abi: params.contract.abi,
+    functionName: params.functionName,
+    args: params?.args,
```

Comment:
Would be nice to have type safety on the args and function name. (such as what I showed you in the PR I made)

Reply:
Agree! I'll add those asap!